### ISSUE TITLE
Debugger dump failed for blessed REF object

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6734]
Migrated from [rt.perl.org#23651](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=23651) (status was 'resolved')

Searchable as RT23651$

**Comment by p5pRT** [comment]
### From @adamkennedy

perlbug didn't work for me\, so I'll do this directly and keep it simple\.

Using perl 5\.8\.0 in the debugger

perl \-de 1

Loading DB routines from perl5db\.pl version 1\.19
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;   DB\<1> x \{\}
0  HASH\(0x804c1b4\)
&nbsp;      empty hash
&nbsp;   DB\<2> x \\\{\}
0  REF\(0x82ed9b8\)
&nbsp;    \-> HASH\(0x82ed9c4\)
&nbsp;         empty hash
&nbsp;   DB\<3> x bless \{\}\, 'Foo'
0  Foo=HASH\(0x82ed9ac\)
&nbsp;      empty hash
&nbsp;   DB\<4> x bless \\\{\}\, 'Foo'
0  Foo=REF\(0x8300e28\)
&nbsp;   DB\<5>


Basically\.\.\. when looking at an object based on a blessed REF reference\, 
the dumping function doesn't descend the REF like it does for an 
unblessed REF reference\.

I figure it's probably a quick and simple fix somewhere in the dumping 
\.pl file\, but I can't pinpoint the exact location\.

Adam K

**Comment by p5pRT** [comment]
@adamkennedy - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@adamkennedy - Status changed from 'open' to 'new'

**Comment by p5pRT** [comment]
### From mcmahon@ibiblio.org


On Friday\, August 29\, 2003\, at 03&#8203;:20 PM\, Adam Kennedy \(via RT\) wrote&#8203;:

> main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
>    DB\<1> x \{\}
> 0  HASH\(0x804c1b4\)
>       empty hash
>    DB\<2> x \\\{\}
> 0  REF\(0x82ed9b8\)
>     \-> HASH\(0x82ed9c4\)
>          empty hash
>    DB\<3> x bless \{\}\, 'Foo'
> 0  Foo=HASH\(0x82ed9ac\)
>       empty hash
>    DB\<4> x bless \\\{\}\, 'Foo'
> 0  Foo=REF\(0x8300e28\)
>    DB\<5>
>
>
> Basically\.\.\. when looking at an object based on a blessed REF 
> reference\,
> the dumping function doesn't descend the REF like it does for an
> unblessed REF reference\.
>
> I figure it's probably a quick and simple fix somewhere in the dumping
> \.pl file\, but I can't pinpoint the exact location\.
>

It was indeed in dumpvar\.pl\, but it was not quick and simple\. There's 
definitely a hole here; there's no function \(as far as I know\) that 
dependably returns the \*base\* class of an object\. I implemented it by 
matching the base class out of stringified references\, though I suppose 
this could break in the future\.

There are a large number of nasty special cases&#8203;: \\undef is a SCALAR 
pointing to undef\, while \\\{\} is a REF pointing to an empty hash \(should 
this be considered a bug\, or is it documented as working that way?\)\.

Output looks like this now&#8203;:

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;   DB\<1> x \{\}
0  HASH\(0x4ae2e4\)
&nbsp;      empty hash
&nbsp;   DB\<2> x \\\{\}
0  REF\(0x4c5134\)
&nbsp;    \-> HASH\(0x4c50bc\)
&nbsp;            empty hash
&nbsp;   DB\<3> x bless \{\}\,"Foo"
0  Foo=HASH\(0x537474\)
&nbsp;      empty hash
&nbsp;   DB\<4> x bless \\\{\}\,"Foo"
0  Foo=REF\(0x1b55b8\)
&nbsp;    \-> HASH\(0x1b5588\)
&nbsp;            empty hash
&nbsp;   DB\<5> x \[\]
0  ARRAY\(0x4b7310\)
&nbsp;      empty array
&nbsp;   DB\<6> x \\\[\]
0  REF\(0x4b7430\)
&nbsp;    \-> ARRAY\(0x4c7760\)
&nbsp;            empty array
&nbsp;   DB\<7> x bless \[\]\,"Bar"
0  Bar=ARRAY\(0x1b557c\)
&nbsp;      empty array
&nbsp;   DB\<8> x bless \\\[\]\,"Bar"
0  Bar=REF\(0x1b566c\)
&nbsp;    \-> ARRAY\(0x35c5dc\)
&nbsp;            empty array
&nbsp;   DB\<9>

&nbsp;  \-\-\- Joe M\.

**Comment by p5pRT** [comment]
### From mcmahon@ibiblio.org

<details><summary>dumpvar.patch</summary>

```diff
--- lib/dumpvar.pl.old	Thu Jul 10 02:12:01 2003
+++ /usr/local/lib/perl5/5.8.1/dumpvar.pl	Tue Sep  2 17:55:41 2003
@@ -136,7 +136,19 @@
       my $val = $v;
       $val = &{'overload::StrVal'}($v) 
 	if %overload:: and defined &{'overload::StrVal'};
-      ($address) = $val =~ /(0x[0-9a-f]+)\)$/ ; 
+      # Match type and address.                      
+      # Unblessed references will look like TYPE(0x...)
+      # Blessed references will look like Class=TYPE(0x...)
+      ($start_part, $val) = split /=/,$val;
+      $val = $start_part unless defined $val;
+      ($item_type, $address) = 
+        $val =~ /([^\(]+)        # Keep stuff that's     
+                                 # not an open paren
+                 \(              # Skip open paren
+                 (0x[0-9a-f]+)   # Save the address
+                 \)              # Skip close paren
+                 $/x;            # Should be at end now
+
       if (!$dumpReused && defined $address) { 
 	$address{$address}++ ;
 	if ( $address{$address} > 1 ) { 
@@ -145,6 +157,7 @@
 	} 
       }
     } elsif (ref \$v eq 'GLOB') {
+      # This is a raw glob. Special handling for that.
       $address = "$v" . "";	# To avoid a bug with globs
       $address{$address}++ ;
       if ( $address{$address} > 1 ) { 
@@ -154,13 +167,15 @@
     }
 
     if (ref $v eq 'Regexp') {
+      # Reformat the regexp to look the standard way.
       my $re = "$v";
       $re =~ s,/,\\/,g;
       print "$sp-> qr/$re/\n";
       return;
     }
 
-    if ( UNIVERSAL::isa($v, 'HASH') ) { 
+    if ( $item_type eq 'HASH' ) { 
+        # Hash ref or hash-based object.
 	@sortKeys = sort keys(%$v) ;
 	undef $more ; 
 	$tHashDepth = $#sortKeys ; 
@@ -193,14 +208,19 @@
 	}
 	print "$sp  empty hash\n" unless @sortKeys;
 	print "$sp$more" if defined $more ;
-    } elsif ( UNIVERSAL::isa($v, 'ARRAY') ) { 
+    } elsif ( $item_type eq 'ARRAY' ) { 
+        # Array ref or array-based object. Also: undef.
+        # See how big the array is.
 	$tArrayDepth = $#{$v} ; 
 	undef $more ; 
+        # Bigger than the max?
 	$tArrayDepth = $#{$v} < $arrayDepth-1 ? $#{$v} : $arrayDepth-1 
 	  if defined $arrayDepth && $arrayDepth ne '';
+        # Yep. Don't show it all.
 	$more = "....\n" if $tArrayDepth < $#{$v} ; 
 	$shortmore = "";
 	$shortmore = " ..." if $tArrayDepth < $#{$v} ;
+
 	if ($compactDump && !grep(ref $_, @{$v})) {
 	  if ($#$v >= 0) {
 	    $short = $sp . "0..$#{$v}  " . 
@@ -220,20 +240,35 @@
 	    return if $DB::signal;
 	    print "$sp$num  ";
 	    if (exists $v->[$num]) {
-	        DumpElem $v->[$num], $s, $m-1;
+                if (defined $v->[$num]) {
+	          DumpElem $v->[$num], $s, $m-1;
+                } 
+                else {
+                  print "undef\n";
+                }
 	    } else {
 	    	print "empty slot\n";
 	    }
 	}
 	print "$sp  empty array\n" unless @$v;
 	print "$sp$more" if defined $more ;  
-    } elsif (  UNIVERSAL::isa($v, 'SCALAR') or ref $v eq 'REF' ) { 
+    } elsif ( $item_type eq 'SCALAR' ) { 
+            unless (defined $$v) {
+              print "$sp-> undef\n";
+              return;
+            }
 	    print "$sp-> ";
 	    DumpElem $$v, $s, $m-1;
-    } elsif ( UNIVERSAL::isa($v, 'CODE') ) { 
+    } elsif ( $item_type eq 'REF' ) { 
+	    print "$sp-> $$v\n";
+            return unless defined $$v;
+	    unwrap($$v, $s+3, $m-1);
+    } elsif ( $item_type eq 'CODE' ) { 
+            # Code object or reference.
 	    print "$sp-> ";
 	    dumpsub (0, $v);
-    } elsif ( UNIVERSAL::isa($v, 'GLOB') ) {
+    } elsif ( $item_type eq 'GLOB' ) {
+      # Glob object or reference.
       print "$sp-> ",&stringify($$v,1),"\n";
       if ($globPrint) {
 	$s += 3;
@@ -242,6 +277,7 @@
 	print( (' ' x ($s+3)) .  "FileHandle({$$v}) => fileno($fileno)\n" );
       }
     } elsif (ref \$v eq 'GLOB') {
+      # Raw glob (again?)
       if ($globPrint) {
        dumpglob($s, "{$v}", $v, 1, $m-1) if $globPrint;
       } elsif (defined ($fileno = fileno(\$v))) {

```
</details>

**Comment by p5pRT** [comment]
### From @adamkennedy

> It was indeed in dumpvar\.pl\, but it was not quick and simple\. There's 
> definitely a hole here; there's no function \(as far as I know\) that 
> dependably returns the \*base\* class of an object\. I implemented it by 
> matching the base class out of stringified references\, though I uppose 
> this could break in the future\.

Well\.\.\. if you are doing a switch workalike\, it's fairly easy to just do
something like\.


my $this = \(reference to check\);

if \( UNIVERSAL&#8203;::isa\( $this\, 'SCALAR' \) \) \{
&nbsp;    \# Treat as a SCALAR\, blessed or otherwise
\} elsif \( UNIVERSAL&#8203;::isa\( $this\, 'ARRAY' \) \) \{
&nbsp;    \# Treat as an ARRAY\, blessed or otherwise
\} elsif \( etc \) \{
&nbsp;    \# etc
\}

The display behaviour with regard to indenting\, does not seem like it is
supposed to be any different regardless of whether the reference is
blessed into an object or not\.

Surely it's easy enough to just use UNIVERSAL&#8203;::isa to switch on their
base types\, rather than parsing their 'ref' strings?

**Comment by p5pRT** [comment]
### From @rgs

Joe's patch was added as change \#21010\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
@jhi - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @jhi

Unfortunately Joe's patch seemed to have broken the 'x' command of the debugger\, so 
I'm retracting the patch and reopening the bug\.

**Comment by p5pRT** [comment]
### From @adamkennedy

> \[rafael \- Thu Sep 04 15&#8203;:46&#8203;:04 2003\]&#8203;:
> 
> Joe's patch was added as change \#21010\.
> 

With this failed and reversed\, you might want to have a look at
Devel&#8203;::Dumpvar\, which does solves this problem correctly\.

**Comment by p5pRT** [comment]
### From @schwern

Joe's patch appears to have gone in\.

$ bleadperl \-de 1

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> x bless \{\}\, "Foo"
0  Foo=HASH\(0x810850\)
&nbsp;     empty hash
&nbsp;  DB\<2> x bless \\\{\}\, "Foo"
0  Foo=REF\(0x8a8ce0\)
&nbsp;   \-> HASH\(0x8a8cf0\)
&nbsp;           empty hash

**Comment by p5pRT** [comment]
@schwern - Status changed from 'open' to 'resolved'
